### ISSUE TITLE
SafeDisc 1 games crash

### ISSUE THREAD

**Issue by NicknineTheEagle** [RibShark/SafeDiscShim#17]
It appears that none of SafeDisc 1 games work for me, ICD always crashes in dplayerx.dll. Tried the following games:
- Need for Speed: High Stakes
- Need for Speed: Porsche Unleashed
- LEGO Racers

Running Windows 10. Using Alcohol 120% to mount Redump disc images. Not an issue with the game code since cracked exe produced by UnSafeDisc works fine. The games do work if Alcohol virtual drive is passed through to Windows XP VM.

Logs:
[LEGORacers.exe_SafeDiscShim.log](https://github.com/RibShark/SafeDiscShim/files/15214047/LEGORacers.exe_SafeDiscShim.log)
[LEGORacers.ICD_SafeDiscShim.log](https://github.com/RibShark/SafeDiscShim/files/15214048/LEGORacers.ICD_SafeDiscShim.log)

Windows event log entries:
```
Faulting application name: LEGORacers.ICD, version: 0.0.0.0, time stamp: 0x37979360
Faulting module name: dplayerx.DLL, version: 0.0.0.0, time stamp: 0x3729dba5
Exception code: 0xc0000005
Fault offset: 0x00013044
Faulting process id: 0x138c
Faulting application start time: 0x01da9f18bcf2fb97
Faulting application path: C:\Games\LEGORacers\LEGORacers.ICD
Faulting module path: D:\dplayerx.DLL
Report Id: 0882d3a8-0083-41f4-baed-b1dc4d246f80
Faulting package full name: 
Faulting package-relative application ID: 
```
```
Faulting application name: LEGORacers.ICD, version: 0.0.0.0, time stamp: 0x37979360
Faulting module name: LEGORacers.ICD, version: 0.0.0.0, time stamp: 0x37979360
Exception code: 0xc0000005
Fault offset: 0x000a867c
Faulting process id: 0x138c
Faulting application start time: 0x01da9f18bcf2fb97
Faulting application path: C:\Games\LEGORacers\LEGORacers.ICD
Faulting module path: C:\Games\LEGORacers\LEGORacers.ICD
Report Id: 89584dd5-b462-46cc-ad53-0f5b8c27a40f
Faulting package full name: 
Faulting package-relative application ID: 
```

**Comment by NicknineTheEagle** [comment]
Figure out the root cause, it looks like the new injection mechanism involving overwriting the entry point does not work SafeDisc 1 ICDs: https://github.com/RibShark/SafeDiscShim/commit/8fedebb00cf7f73a8514f0bff292dd3a6f7a2d1e
Simply calling `Initialize` from `DllMain` like in older revisions fixes the crash.

**Comment by NicknineTheEagle** [comment]
While initializing from DllMain is unsafe due to loader lock, as far as I can tell, there's literally no other way to do this for ICD. dplayerx.dll seems to be loaded before ICD entry point (import table?) and it needs Secdrv driver in its own DllMain meaning that the shim must be initialized by this point or dplayerx.dll will crash.

**Comment by mirh** [comment]
My memory aches trying to remember what rule of thumb I must have read in the past, but you might want to take a notice at:
https://www.vogons.org/viewtopic.php?t=51818
https://github.com/ElliotKillick/LdrLockLiberator

**Comment by All3n-Hunt3r** [comment]
Stumbled upon this thread because I'm troubleshooting Mech 4 Vengeance which is a SafeDisc 1 game. Event Viewer seems to be reporting both MW4.ICD and dplayerx.dll go faulty and I'm guessing the latter is the culprit.

  I have directplay installed and all that, but how do I go about fixing this issue?

**Comment by NicknineTheEagle** [comment]
> Stumbled upon this thread because I'm troubleshooting Mech 4 Vengeance which is a SafeDisc 1 game. Event Viewer seems to be reporting both MW4.ICD and dplayerx.dll go faulty and I'm guessing the latter is the culprit.
> 
> I have directplay installed and all that, but how do I go about fixing this issue?

dplayerx.dll has nothing to do with dplay.dll or dplayx.dll, it's a SafeDisc dll. The real issue is a bug in SafeDiscShim as described above.

**Comment by NicknineTheEagle** [comment]
> Ahh, I understand. Is there anything I can do to fix this issue myself?

Nope. For now, you can try using UnSafeDisc to remove SafeDisc from the game.

> perhaps I could try uninstalling SafeDiscShim and reinstalling the previous version?

No, I was referring to prerelease code, there is no old release to revert to.

**Comment by All3n-Hunt3r** [comment]
Ah, gotcha. Does unsafedisc only affect the game install or does it mess with the disc files too?

**Comment by NicknineTheEagle** [comment]
I don't see how it would mess with the disc files seeing how you can't just write to discs. It just removes SafeDisc from game exe.

**Comment by All3n-Hunt3r** [comment]
Thanks for the help then! I hope this’ll work, and that before long Shim gets updated

**Comment by All3n-Hunt3r** [comment]
Goddamnit, UnSafeDisc keeps throwing up a useless exception occurred error I don't know how to fix. So either I'll have to re-install and install the 2.0 patch (instead of 3.0) to use the nocd executable, wait for SafeDiscShim to get updated at some point or....I dunno.

**Comment by RibShark** [comment]
Should be fixed as of v0.9.0. If it is still broken, please let me know.

**Comment by All3n-Hunt3r** [comment]
Mech 4 still does not seem to work with the latest version of SafeDiscShim. And yes, I restarted my computer.
